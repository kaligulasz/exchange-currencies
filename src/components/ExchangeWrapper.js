import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Sync } from 'styled-icons/octicons/Sync';
import currencyJs from 'currency.js';
import ReactRouterPropTypes from 'react-router-prop-types';

// Actions
import { syncWithMainPocket, setCurrencyAmount } from '../actions/exchangePocketActions';
import { updateMainPocket } from '../actions/mainPocketActions';

// Reducers
import { getMainPocket, getMainPocketCurrencyList } from '../reducers/mainPocketReducer';
import { getExchangePocket } from '../reducers/exchangePocketReducer';
import { getCurrencyRates } from '../reducers/currencyRatesReducer';
import ExchangeItem from './ExchangeItem';

// Components
import Button from './Button';
import ExchangeNavigation from './ExchangeNavigation';

// Styled
const PrimaryWrapper = styled.section`
  height: ${props => props.theme.primaryHeight};
  min-height: ${props => props.theme.primaryMinHeight};
  
  @media (max-width: ${props => props.theme.mediaQuery.tablet}) and (orientation: landscape) {
    height: ${props => props.theme.mobileHeight};
  }
`;

const ItemsWrapper = styled.section`
  background: ${props => props.theme.color.primaryGradient};
  height: 73%;
`;

const SyncIcon = styled(Sync)`
  display: block;
  float: right;
  margin-left: 0.3rem;
`;

const Footer = styled.footer`
  text-align: center;
  background: linear-gradient(90deg, rgba(18,67,111,1) 0%, rgba(47,144,114,1) 100%);
  display: flex;
  justify-content: center;
  align-items: center;
  height: 15%;
`;

const ExchangeWrapper = ({
  mainPocket,
  match,
  history,
  syncWithMainPocketAction,
  exchangePocket,
  currencyRates,
  setCurrencyAmountAction,
  updateMainPocketAction,
  currencyList,
}) => {
  const { changingFromCurrency = 'usd' } = match.params;
  const { changingToCurrency = 'eur' } = match.params;

  useEffect(() => {
    syncWithMainPocketAction(mainPocket);
  }, []);

  const changingFromPocket = exchangePocket[changingFromCurrency];
  const changingToPocket = exchangePocket[changingToCurrency];

  const actualRate = () => (
    currencyRates[changingToCurrency].rate / currencyRates[changingFromCurrency].rate
  );

  const onAmountChange = (value) => {
    const amountInDollars = currencyJs(value) / currencyRates[changingFromCurrency].rate;
    const selectedCurrencyNewAmount = currencyJs(
      mainPocket[changingFromCurrency].amount - currencyJs(value),
    ).value;

    const targetCurrencyNewAmount = (
      currencyJs(amountInDollars) * currencyRates[changingToCurrency].rate
      + currencyJs(mainPocket[changingToCurrency].amount)
    ).value;

    setCurrencyAmountAction(changingFromCurrency, selectedCurrencyNewAmount);
    setCurrencyAmountAction(changingToCurrency, targetCurrencyNewAmount);
  };

  const handleOnExchange = () => {
    if (changingToPocket.amount !== mainPocket[changingToCurrency].amount) {
      updateMainPocketAction(exchangePocket);

      history.push('/');
    }
  };

  return (
    <PrimaryWrapper>
      <ExchangeNavigation currencyList={currencyList} />

      <ItemsWrapper>
        <ExchangeItem
          mainPocket={mainPocket}
          pocket={changingFromPocket}
          onAmountChange={onAmountChange}
          primary
        />

        <ExchangeItem
          mainPocket={mainPocket}
          pocket={changingToPocket}
          actualRate={actualRate()}
          changingFromCurrencySymbol={changingFromPocket.currencySymbol}
          secondary
        />
      </ItemsWrapper>

      <Footer>
        <Button
          onClick={handleOnExchange}
          primary
          uppercase
        >
          exchange
          <SyncIcon size="19" />
        </Button>
      </Footer>
    </PrimaryWrapper>
  );
};

ExchangeWrapper.propTypes = {
  mainPocket: PropTypes.object.isRequired,
  match: ReactRouterPropTypes.match.isRequired,
  syncWithMainPocketAction: PropTypes.func.isRequired,
  exchangePocket: PropTypes.object.isRequired,
  currencyRates: PropTypes.object.isRequired,
  setCurrencyAmountAction: PropTypes.func.isRequired,
  updateMainPocketAction: PropTypes.func.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
  currencyList: PropTypes.array.isRequired,
};

const mapStateToProps = state => ({
  mainPocket: getMainPocket(state),
  exchangePocket: getExchangePocket(state),
  currencyRates: getCurrencyRates(state),
  currencyList: getMainPocketCurrencyList(state),
});

export default connect(
  mapStateToProps,
  {
    syncWithMainPocketAction: syncWithMainPocket,
    setCurrencyAmountAction: setCurrencyAmount,
    updateMainPocketAction: updateMainPocket,
  },
)(withRouter(ExchangeWrapper));
