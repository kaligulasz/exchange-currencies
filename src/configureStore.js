import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import reducers from './reducers/index';
import rootSaga from './sagas/index';

/**
 * Configure Store
 * @return {Object} - The whole state tree the your application
 */
const configureStore = () => {
  let store;
  const sagaMiddleware = createSagaMiddleware();
  const middlewares = [
    sagaMiddleware,
  ];

  if (process.env.NODE_ENV === 'development') {
    const composeEnhancers = composeWithDevTools({ name: 'exchange-currencies' });
    store = createStore(reducers, composeEnhancers(applyMiddleware(...middlewares)));
  } else {
    store = applyMiddleware(...middlewares)(createStore)(reducers);
  }

  sagaMiddleware.run(rootSaga);

  return store;
};

export default configureStore;
