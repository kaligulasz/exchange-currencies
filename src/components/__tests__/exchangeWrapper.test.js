import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import ExchangeWrapper from '../ExchangeWrapper';

const mockStore = configureStore();

let store;
const initialStore = {
  mainPocket: {
    pockets: {
      usd: {
        currencySymbol: '$',
        currency: 'usd',
        description: 'USD - American Dollar',
        amount: 0.73,
        id: 1,
      },
      eur: {
        currencySymbol: '€',
        currency: 'eur',
        description: 'EUR - Euro',
        amount: 46.38,
        id: 2,
      },
      gbp: {
        currencySymbol: '£',
        currency: 'gbp',
        description: 'GBP - British Pound',
        amount: 25.63,
        id: 3,
      },
    },
  },
  currencyRates: {
    status: 'initial',
    currencies: {
      usd: {
        currency: '$',
        rate: 1,
        id: 1,
      },
      eur: {
        currency: '€',
        rate: 0.89194,
        id: 2,
      },
      gbp: {
        currency: '£',
        rate: 0.75913,
        id: 3,
      },
    },
  },
  exchangePocket: {
    pockets: {
      usd: {
        currencySymbol: '$',
        currency: 'usd',
        description: 'USD - American Dollar',
        amount: 0.73,
        id: 1,
      },
      eur: {
        currencySymbol: '€',
        currency: 'eur',
        description: 'EUR - Euro',
        amount: 46.38,
        id: 2,
      },
      gbp: {
        currencySymbol: '£',
        currency: 'gbp',
        description: 'GBP - British Pound',
        amount: 25.63,
        id: 3,
      },
    },
  },
};

beforeEach(() => {
  store = mockStore(initialStore);
});

describe('<ExchangeWrapper />', () => {
  it('renders <ExchangeWrapper />', () => {
    const component = shallow(
      <ExchangeWrapper store={store} />,
    ).dive();

    expect(component).toMatchSnapshot();
  });
});
