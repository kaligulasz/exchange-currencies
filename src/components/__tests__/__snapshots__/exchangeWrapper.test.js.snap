// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExchangeWrapper /> renders <ExchangeWrapper /> 1`] = `
<withRouter(ExchangeWrapper)
  currencyList={
    Array [
      "usd",
      "eur",
      "gbp",
    ]
  }
  currencyRates={
    Object {
      "eur": Object {
        "currency": "€",
        "id": 2,
        "rate": 0.89194,
      },
      "gbp": Object {
        "currency": "£",
        "id": 3,
        "rate": 0.75913,
      },
      "usd": Object {
        "currency": "$",
        "id": 1,
        "rate": 1,
      },
    }
  }
  exchangePocket={
    Object {
      "eur": Object {
        "amount": 46.38,
        "currency": "eur",
        "currencySymbol": "€",
        "description": "EUR - Euro",
        "id": 2,
      },
      "gbp": Object {
        "amount": 25.63,
        "currency": "gbp",
        "currencySymbol": "£",
        "description": "GBP - British Pound",
        "id": 3,
      },
      "usd": Object {
        "amount": 0.73,
        "currency": "usd",
        "currencySymbol": "$",
        "description": "USD - American Dollar",
        "id": 1,
      },
    }
  }
  mainPocket={
    Object {
      "eur": Object {
        "amount": 46.38,
        "currency": "eur",
        "currencySymbol": "€",
        "description": "EUR - Euro",
        "id": 2,
      },
      "gbp": Object {
        "amount": 25.63,
        "currency": "gbp",
        "currencySymbol": "£",
        "description": "GBP - British Pound",
        "id": 3,
      },
      "usd": Object {
        "amount": 0.73,
        "currency": "usd",
        "currencySymbol": "$",
        "description": "USD - American Dollar",
        "id": 1,
      },
    }
  }
  setCurrencyAmountAction={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  syncWithMainPocketAction={[Function]}
  updateMainPocketAction={[Function]}
/>
`;
